app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2019 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.arpnetworking.metrics.portal.reports;
+
+import com.google.common.base.CaseFormat;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Maps;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.name.Names;
+import models.internal.impl.ChromeScreenshotReportSource;
+import models.internal.impl.HtmlReportFormat;
+import models.internal.impl.PdfReportFormat;
+import models.internal.reports.Recipient;
+import models.internal.reports.Report;
+import models.internal.reports.ReportFormat;
+import models.internal.reports.ReportSource;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Utilities for execution of {@link Report}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class ReportExecution {
+
+    /**
+     * Render and send a report.
+     *
+     * @param report The report to render and execute.
+     * @param injector A Guice injector to pull dependencies out of.
+     * @param scheduled The instant the report was scheduled for.
+     * @return A CompletionStage that completes when sending has completed for every recipient.
+     */
+    public static CompletionStage<Report.Result> execute(final Report report, final Injector injector, final Instant scheduled) {
+        verifyDependencies(report, injector);
```

Comment:
This seems unnecessary. It will either pass or fail every time. Imo we should not add defensive runtime code on each execution for what is a configuration problem. I expect the job will fail gracefully if it cannot load a renderer or sender when it's actually executing. I assume this is more of a all or none type failure protection, I'm not particularly concerned because it's config based not data based.

Reply:
Without this check, `execute()` throws synchronously if it can't find all the Renderers it needs, and asynchronously if it can't find all the Senders it needs. The mixture of [a]synchronous throwing seemed gross to me, and I feel that a synchronous throw is more appropriate here than asynchronous (since the impending failure _can_ be determined quickly).

What counts as a "graceful" failure here? Aborting and logging an exception?

Reply:
If you want to throw synchronously and want to force users to deal with it then this should probably be a checked exception.

However, because there is a completion stage returned user's will need to deal with its failure no matter what. If you want to check up front, why not just return a failed completion stage? (Why throw and return a failable future?)

Reply:
Good point! Fixed: it now only ever throws asynchronously.